### PR TITLE
sz: Support S2 (close #429)

### LINKED ISSUE #429: Add S2 extension of snappy compression
## Introduce the format you are requesting.
[`S2`](https://github.com/klauspost/compress/tree/master/s2) extension of `snappy` compression

## What do YOU use this format for?
Because it has benefits over Snappy:
- Better compression
- Adjustable compression (3 levels)
- Concurrent stream compression
- Faster decompression, even for Snappy compatible content
- Concurrent Snappy/S2 stream decompression
- Skip forward in compressed stream
- Random seeking with indexes
- Compatible with reading Snappy compressed content
- Smaller block size overhead on incompressible blocks
- Block concatenation
- Block Dictionary support
- Uncompressed stream mode
- Automatic stream size padding
- Snappy compatible block compression

## What is the format's conventional file extension(s)?

To match existing utility `s2c`/`s2d` is should be `.s2` file extension.

## Please link to the format's formal or official specification(s).

https://github.com/klauspost/compress/tree/master/s2

## Which Go libraries could be used to implement this format?

https://github.com/klauspost/compress/tree/master/s2

### PR DESCRIPTION
**Pull request by mholt**
@CompuRoot maybe you can take a look at this. I'll likely just merge this and then we can change it later if needed.

@klauspost -- Thank you for yet again, another great package! (One minor request: It could be useful if the compression levels were exported.)

### CODE REVIEW

**Comment by klauspost** [comment]
For now here is a teaser :)

image

**Comment by mholt** [comment]
Wow! So, do you think I should wait for that?

File: sz.go [hunk 1]
```diff
@@ -36,13 +65,68 @@ func (sz Sz) Match(_ context.Context, filename string, stream io.Reader) (MatchR
...
+	if sz.S2.IgnoreCRC {
+		opts = append(opts, s2.ReaderIgnoreCRC())
+	}
+	if sz.S2.IgnoreStreamIdentifier {
+		opts = append(opts, s2.ReaderIgnoreStreamIdentifier())
+	}
+	if sz.S2.MaxBlockSize != 0 {
+		opts = append(opts, s2.ReaderMaxBlockSize(sz.S2.MaxBlockSize))
+	}
+	return io.NopCloser(s2.NewReader(r, opts...)), nil
```
Comment (klauspost) [inline sz.go:115]: A bit sad the API is truncated to an `io.ReadCloser`, but I guess the users can flip back to the org package for more extensive API.

**Review by CompuRoot — APPROVED** [review]
Thank you for picking it so quickly !

The only suggestion is to change file extension `.s2` to `.sz2` or `.s2z` to be on pair with other and to reflect relation to `Snappy`

**Comment by CompuRoot** [comment]
> For now here is a teaser :)

And bellow is result of testing a small (140Mb) `.vdi` file with [`kopia`](https://github.com/kopia/kopia):

```
     Compression                Compressed   Throughput   Allocs   Usage
------------------------------------------------------------------------------------------------
  0. s2-parallel-8              62.9 MB      1.5 GB/s     1023     197 MB
  1. s2-default                 62.9 MB      1.1 GB/s     1050     195.9 MB
  2. s2-parallel-4              62.9 MB      1 GB/s       995      187.5 MB
  3. pgzip-best-speed           58.9 MB      816.1 MB/s   1352     214.1 MB
  4. s2-better                  60.6 MB      739.4 MB/s   1019     182.4 MB
  5. pgzip                      56.9 MB      474.1 MB/s   1294     210.9 MB
  6. lz4                        64.7 MB      433.7 MB/s   27       204.1 MB (deprecated)
  7. zstd-fastest               54.4 MB      352.6 MB/s   6311     212.4 MB
  8. deflate-best-speed         58.9 MB      322.3 MB/s   38       135 MB
  9. zstd                       52.4 MB      322.3 MB/s   3266     156.3 MB
 10. deflate-default            56.9 MB      186.6 MB/s   38       135.3 MB
 11. gzip-best-speed            58.4 MB      185.6 MB/s   44       135.4 MB
 12. zstd-better-compression    51.6 MB      108.9 MB/s   3222     176 MB
 13. gzip                       55.9 MB      47.3 MB/s    44       135 MB
 14. zstd-best-compression      47 MB        31.8 MB/s    3367     207.5 MB (deprecated)
 15. pgzip-best-compression     56.1 MB      31.2 MB/s    1299     171.6 MB
 16. deflate-best-compression   56.1 MB      8.6 MB/s     39       135.4 MB
 17. gzip-best-compression      55.9 MB      7.3 MB/s     43       135 MB
```

File: sz.go [hunk 2]
```diff
@@ -36,13 +65,68 @@ func (sz Sz) Match(_ context.Context, filename string, stream io.Reader) (MatchR
...
+// EXPERIMENTAL: May be changed or removed without a major version bump.
+type S2Level int
+
+// Compression levels for S2.
+// EXPERIMENTAL: May be changed or removed without a major version bump.
+const (
+	S2LevelNone   S2Level = 1
+	S2LevelFast   S2Level = 2
+	S2LevelBetter S2Level = 3
+	S2LevelBest   S2Level = 4
```
[the remark this answers is not in the document]
Comment (mholt) [reply sz.go:128]: Thanks! I was initially going off the unexported values but I'll do this instead.

**Comment by CompuRoot** [comment]
> @CompuRoot I am not religious about it, but I have used `.s2` for my [commandline tool](https://github.com/klauspost/compress/blob/master/s2/cmd/s2c/main.go#L55).
> 
> The snappy relation is not too important, except for backwards compatibility of reading snappy data.

You right, if there already exists CLI tool with this extension, then keeping `.s2` extension for compatibility is a better solution.

**Pull request merged**